sark/tseries.py

```diff
@@ -0,0 +1,171 @@
+    specific to this function are ignored, and all other keyword options are
+    passed on to the backend transparently; it's a noop.
+
+    Parameters
+    ----------
+    fpath : Union[str, Path]
+        Path to the dataset file
+
+    date_cols : List[int, str] (for "multicol" mode)
+        List of columns to be combined to construct the datetime values
+
+    col_units : str (for "table" mode)
+        Time units for the columns
+
+    zero_idx : bool (for "table" mode)
+        Whether the columns are zero indexed.  When the columns represent
+        hours, or minutes, it is common to number them as nth hour.  Which
+        means they are counted starting at 1 instead of 0.  Set this to False
+        if that is the case.
+
+    **kwards : Dict
+        Other keyword arguments passed on to the reader backend.  Any options
+        passed here takes precedence, and overwrites other values inferred from
+        the earlier keyword arguments.
+
+    Returns
+    -------
+    ts : Series/DataFrame
+        The time series is returned as a series or a dataframe depending on the
+        number of other columns that are present.
+
+    Examples
+    --------
+
+    To skip specific rows, maybe because they have bad data, or are empty, you
+    may use the `skiprows` option.  It can be set to a list-like where the
+    entries are row indices (numbers).
+
+    >>> read_timeseries("mydata.csv", source_t="table", col_units="hour",
+    ...     skiprows=range(1522, 5480))  # doctest: +SKIP
+
+    The above example skips rows 1522-5480.
+
+    Similarly, data type of the column values can be controlled by using the
+    `dtype` option.  When set to a `numpy.dtype`, all values will be read as
+    that type, which is probably relevant for the "table" mode.  In the
+    "multicol" mode, the types of the values can be controlled at the column
+    level by setting it to a dictionary, where the key matches a column name,
+    and the value is a valid `numpy.dtype`.
+
+    """
+    # FIXME: parse_dates & index_col assumes input is oriented as portrait
+    if source_t == "table":
+        if col_units is None:
+            raise ValueError("col_units: missing time unit for columns")
+        ts = from_table(
+            fpath, col_units=col_units, zero_idx=zero_idx, **kwargs,
+        )
+    elif source_t == "multicol":
+        if date_cols is None:
+            raise ValueError("date_cols: missing list of datetime columns")
+        ts = from_multicol(fpath, date_cols=date_cols, **kwargs,)
+    else:
+        if source_t:
+            warn(f"{source_t}: unsupported source, falling back to default")
+        ts = pd.read_csv(fpath, **kwargs)
+    return ts
+
+
+def from_table(fpath: _path_t, *, col_units: str, zero_idx: bool, **kwargs):
+    """Read a time series from a tabular file.
+
+    See Also
+    --------
+    read_timeseries : see for full documentation, main entrypoint for users
+
+    """
+    assert col_units in _time_units
```

Comment:
Is it possible to avoid hardcoding this? I.e. just relying on allowed Pandas units; if a user gets it wrong, 'to_timedelta' would just break. This leaves it open to be e.g. years in index and weeks in columns, etc.

Reply:
I thought about this, couldn't really come to a conclusion, before resorting to limiting the options :-p. I'll experiment a bit more and see what works best.

Reply:
I thought (tested) some more, and decided to be very restrictive, and stick to examples that I have (hourly and monthly). I'll consider making it more generic if users ask for it.  The other possibilities are very complex, because the starting of the periods are may not match, as in, a year/month can start midweek.